agent/agent_endpoint.go

```diff
@@ -1487,3 +1490,90 @@ func (s *HTTPServer) AgentHost(resp http.ResponseWriter, req *http.Request) (int
 
 	return debug.CollectHostInfo(), nil
 }
+
+// AgentBootstrapGossipKey
+//
+// PUT /v1/agent/bootstrap-gossip-key
+//
+// Allows delivering the gossip encryption key via API to avoid it being present
+// on disk if -encrypt-key-from-api and -disable-keyring-file are used.
+func (s *HTTPServer) AgentBootstrapGossipKey(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	// Fetch the ACL token, if any, and enforce agent policy. We require
+	// AgentWrite since we are effectively allowing the agent to join the cluster
+	// by delivering it's key.
+	//
+	// TODO(banks): is this necessary? Seems safer to be more restrictive rather
+	// than less for now. It seems the worst an attacker could do if unprotected
+	// would be to force the agent to join some other gossip pool they have setup
+	// but that could be a bad thing if there are multiple Consul clusters in the
+	// DC for example.
+	var token string
+	s.parseToken(req, &token)
+	rule, err := s.agent.resolveToken(token)
+	if err != nil {
+		return nil, err
+	}
+	if rule != nil && !rule.AgentWrite(s.agent.config.NodeName) {
+		return nil, acl.ErrPermissionDenied
+	}
+
+	var args api.AgentBootstrapGossipKey
+	if err := decodeBody(req, &args, nil); err != nil {
+		return nil, &BadRequestError{Reason: fmt.Sprintf("Request decode failed: %v", err)}
+	}
+
+	// Decode key - should be base64
+	key, err := base64.StdEncoding.DecodeString(args.SecretKey)
+	if err != nil {
+		return nil, &BadRequestError{Reason: fmt.Sprintf("SecretKey should be base64 encoded bytes: %v", err)}
+	}
+	// Note that the keyring update will error below if we don't do this, but this
+	// way we get to return 400 rather than 500.
+	if err := memberlist.ValidateKey(key); err != nil {
+		return nil, &BadRequestError{Reason: fmt.Sprintf("SecretKey invalid: %v", err)}
+	}
+
+	// All good, set the key directly
+	err = s.agent.SetGossipKey(key)
+	if err == ErrGossipBootstrapped {
+		resp.WriteHeader(http.StatusForbidden)
+		fmt.Fprint(resp, err.Error())
+		return nil, nil
+	}
```

Comment:
trivial: seems like a place for `else if`

Reply:
Yeah. I considered that, I'm in two minds if it's any better especially if you are returning in the first case anyway. This way if the first case is ever changed to not return by mistake we'd still handle the error here so seemed marginally safer 🤷‍♂️ . Happy to change if others have any opinion though.